### ISSUE TITLE
PHP errors

### ISSUE THREAD

**Issue by cbrody** [backlineint/ui_patterns_pattern_lab#17]
Hi, I'm having some problems using this module. Visiting /patterns results in an error page beginning:

`The website encountered an unexpected error. Please try again later.</br></br><em class="placeholder">InvalidArgumentException</em>: The URI &#039;&#039; is invalid. You must use a valid URI scheme. in <em class="placeholder">Drupal\Core\Url::fromUri()</em> (line <em class="placeholder">281</em> of <em class="placeholder">core/lib/Drupal/Core/Url.php</em>). <pre class="backtrace">Drupal\Core\Template\TwigExtension-&gt;getLink(NULL, NULL, Array) (Line: 64)`

Visiting other content pages results in the following notices:

`Notice: Only variables should be passed by reference in Drupal\ui_patterns_pattern_lab\Plugin\Deriver\PatternLabDeriver->templateExists() (line 147 of modules/contrib/ui_patterns_pattern_lab/src/Plugin/Deriver/PatternLabDeriver.php).`

I'm using Acquia Dev Desktop with php v 7.0.14.

**Comment by backlineint** [comment]
Hi Chris,

What Pattern Lab instance are you using the module with? Is it something publicly available that I could pull down and try myself, or is it a private repo?

That said, the first error looks similar to something that I saw while testing with the Emulsify theme in #1 For emulsify the error is being caused by /components/_patterns/01-atoms/09-menu/tab/tab.twig

```
<li{{ attributes }}>
  {# Detect if Drupal #}
  {% if attributes %}
    {{ link(link['#title'], link['#url'], { 'class' : link_classes }) }}
  {% else %}
    <a href="{{ tab_link }}" class="tabs__link tabs__link--local-tasks{% if key == 0 %} is-active{% endif %}">{{ tab_text }}</a>
  {% endif %}
</li>

```
I think the patterns page passes the if attributes check, but then the pattern doesn't have preview data for the link array, causing the error. If this is the cause in your case, I think it should at least be possible to add the necessary data to the pattern definition file so that the link can at least be rendered.

For debugging purposes it is possible to ignore a pattern by adding the following to the pattern definition file in pattern lab (more info in the docs [here](https://www.drupal.org/docs/8/modules/ui-patterns-pattern-lab/overriding-automatic-pattern-discovery)):

`ui_pattern_definition:
  ignore: true`

So if we think we have it narrowed down to a particular pattern we can ignore it and see if the pattern library page will load without it.

As for the notice, I can dupe that one locally so I should be able to clean it up.

Thanks for submitting this issue - the more issues we can identify and fix by running this with a variety of pattern lab instances the more useful it should end up being for folks.

**Comment by backlineint** [comment]
I was able to duplicate the URI error with a pattern that uses:

`{{ link(link['#title'], link['#url'], { 'class' : link_classes }) }}`

But doesn't provide data for the link array.  If I add the following data in the pattern definition file it resolves the error:

```
link:
  '#title': "Title"
  '#url': "http://yahoo.com"
link_classes: "myclass"

```

In PR #19 I also cleaned up the warnings you mentioned. That has been merged into the latest dev version of the module.

**Comment by danny-englander** [comment]
I've run into the same issue as well (using the latest dev of this module) using Emulsify when I visit the `/patterns` page.

**Comment by danny-englander** [comment]
Oh I see, the error might be caused by ` /components/_patterns/01-atoms/09-menu/tab/tab.twig` for which I need to ignore I am guessing in my `MY_THEME.ui_patterns.yml` file?

**Comment by backlineint** [comment]
Hi @danny-englander - you're correct that this is being caused by /components/_patterns/01-atoms/09-menu/tab/tab.twig in Emulisfy. You should be able to work around it by adding one of the following in /components/_patterns/01-atoms/09-menu/tab/tab.yml

```
link:
  '#title': "Title"
  '#url': "http://yahoo.com"
link_classes: "myclass"
```

This example data should be enough for UI Patterns to be able to render a preview of the pattern.

Alternatively, you could add 

```
ui_pattern_definition: 
  ignore: true
```

Which will cause UI Patterns Pattern lab to ignore the pattern.

Testing this module with Emulsify seems like a common starting point for folks, so I'm going to try to do a little more to improve this.  More documentation might help, and I also wonder if I could submit a PR against Emulsify that would maintain Emulsify functionality but prevent this hiccup.

There is also a [UI Patterns PR around handling missing patterns more gracefully](https://github.com/nuvoleweb/ui_patterns/issues/196). I'm going to take a closer look at that issue - I  wonder if that or a similar fix could help improve this situation as well.

**Comment by danny-englander** [comment]
@backlineint yes, I see now, either of those methods work and I can now see my patterns. Thank you!
